src/simulators/circuit_executor.hpp

```diff
@@ -0,0 +1,1174 @@
+
+    // Pass through circuit header and add metadata
+    result.header = circ.header;
+    result.shots = circ.shots;
+    result.seed = circ.seed;
+    result.metadata.add(parallel_shots_, "parallel_shots");
+    result.metadata.add(parallel_state_update_, "parallel_state_update");
+#ifdef AER_CUSTATEVEC
+    if (sim_device_ == Device::GPU)
+      result.metadata.add(cuStateVec_enable_, "cuStateVec_enable");
+#endif
+
+    // Add timer data
+    auto timer_stop = myclock_t::now(); // stop timer
+    double time_taken =
+        std::chrono::duration<double>(timer_stop - timer_start).count();
+    result.time_taken = time_taken;
+  }
+  // If an exception occurs during execution, catch it and pass it to the output
+  catch (std::exception &e) {
+    result.status = ExperimentResult::Status::error;
+    result.message = e.what();
+  }
+}
+
+template <class state_t>
+void Executor<state_t>::run_with_sampling(const Circuit &circ, state_t &state,
+                                          ExperimentResult &result,
+                                          RngEngine &rng, const uint_t shots) {
+  auto &ops = circ.ops;
+  auto first_meas = circ.first_measure_pos; // Position of first measurement op
+  bool final_ops = (first_meas == ops.size());
+
+  // allocate qubit register
+  state.enable_cuStateVec(cuStateVec_enable_);
+  state.allocate(circ.num_qubits, circ.num_qubits);
+  state.set_num_global_qubits(circ.num_qubits);
+  state.enable_density_matrix(!has_statevector_ops_);
+
+  // Run circuit instructions before first measure
+  state.initialize_qreg(circ.num_qubits);
+  state.initialize_creg(circ.num_memory, circ.num_registers);
+
+  state.apply_ops(ops.cbegin(), ops.cbegin() + first_meas, result, rng,
+                  final_ops);
+
+  // Get measurement operations and set of measured qubits
+  measure_sampler(circ.ops.begin() + first_meas, circ.ops.end(), shots, state,
+                  result, rng);
+}
+
+template <class state_t>
+void Executor<state_t>::run_circuit_with_sampling(Circuit &circ,
+                                                  const Config &config,
+                                                  ExperimentResult &result) {
+  state_t state;
+
+  // Optimize circuit
+  Noise::NoiseModel dummy_noise;
+
+  auto fusion_pass = transpile_fusion(circ.opset(), config);
+  fusion_pass.optimize_circuit(circ, dummy_noise, state.opset(), result);
+
+  auto max_bits = get_max_matrix_qubits(circ);
+
+  // Implement measure sampler
+  if (parallel_shots_ <= 1) {
+    // Set state config
+    state.set_config(config);
+    state.set_parallelization(parallel_state_update_);
+    state.set_global_phase(circ.global_phase_angle);
+
+    state.set_distribution(1);
+    state.set_max_matrix_qubits(max_bits);
+
+    RngEngine rng;
+    rng.set_seed(circ.seed);
+    run_with_sampling(circ, state, result, rng, circ.shots);
+  } else {
+    // Vector to store parallel thread output data
```

Comment:
I'm confused that `run_circuit_with_sampling`  needs to consider about `parallel_shots_ `. If we use `sampling`, `state` will be calculated once and then get bitstrings from the `state`.

Reply:
removed unnecessary functions and codes